clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java

```diff
@@ -317,6 +370,10 @@ public void close() {}
         if (!isCompressed())
             return uncompressedIterator();
 
+        if (isSimplified()) {
```

Comment:
I think we can even further optimize it here (assuming we keep this `iterator` as is and create a new iterator constructor class, see my meta comment). Then in the new iterator:

1. still check if `isCompressed`, if not, use this `simplifiedIterator`;
2. otherwise, we can have an even simpler iterator which directly skip the uncompressed bytes for key / value.